### ISSUE TITLE
Concurrent::RejectedExecutionError when running with minitest/autorun via JRuby

### ISSUE THREAD

**Issue by iNecas** [ruby-concurrency/concurrent-ruby#248]
I've been hitting Concurrent::RejectedExecutionError when running my test-stuite. It was ok with MRI, but switching to JRuby was the game-changer. It turns out there seems to be an at_exit trap that shuts down the Java thread pool. I've first tried to fix that using the same approach as in this test suite, reinitializing the Concurrent.configuration. However, there still was an issue with `@root` variable in actor keeping the reference to an old thread pool via Delayed (https://github.com/ruby-concurrency/concurrent-ruby/blob/124ea7e5de7477747ebc149bb4b9c34e22457995/lib/concurrent/actor.rb#L42-L46)

So I needed to reinitialize this as well the hackish way:

https://github.com/iNecas/dynflow/commit/b9a7bf2345404e35dda26bf8ecb86f0cf92fcc8c

What do you think about having set of helpers that helps with setting the testing environment and this kind of issues (I guess there might be more: we're not using all the concurrent-ruby features.)

**Comment by jdantonio** [comment]
@iNecas We definitely need to figure out a better way to handle shutting down the Java thread pools.

The reason for the `at_exit` hook is that Java's thread pools will prevent the JVM from shutting down if the thread pool is not explicitly shut down. Under JRuby our thread pools are simple wrappers around the thread pools from `java.util.concurrent` and it's the underlying thread pool that causes the problem. When I first encountered this it took me a while to figure out what was happening so I grew concerned that users would have the same problem. The `at_exit` hook solved the problem by forcing the pool to be explicitly shut down. Our current `Actor` implementation came much later and that's when the `at_exit` hook became a problem.

I wonder if we're trying to solve this problem the wrong way. Rather than add extra layers *around* the thread pools, could we solve the problem allowing the creator of a given pool to *not* create the `at_exit` hook? Something like:

```ruby
pool = Concurrent::FixedThreadPoool.new(5, set_shutdown_hook: false)
```

The new option would be ignored unless running under JRuby and the default would be `true` to ensure proper shutdown. It would then be a simple matter of updating the various calls to `set_shutdown_hook` (such as [here](https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/executor/java_thread_pool_executor.rb#L78) to honor the true/false flag. The `Actor` framework could then create it's internal thread pool without the `at_exit` hook and then take responsibility for making sure the pool is shutdown properly. (And, of course, we would have to thoroughly document the ramifications of not enabling the shutdown hook.)

@pitr-ch Thoughts?

**Comment by pitr-ch** [comment]
It's also related to #227.

Thanks @iNecas for filling. (I know you did not have too since you've already told me over coffee.)

Thanks @jdantonio for the info, I did not know about https://github.com/ruby-concurrency/concurrent-ruby/search?q=set_shutdown_hook&type=Code&utf8=%E2%9C%93 

I think the right approach could be: not installing any `at_exit` hooks by default to avoid any hacky re-initializations when something is terminated prematurely. The global thread-pools should be there available always. User created thread-pools are up to user to terminate properly, I like `set_shutdown_hook:` option (it should be honored on MRI too I think, better to shutdown properly then let MRI to kill the threads in random order [I've already saw issues with that.]). Still for convenience there should be something like `Concurrent.auto_terminate!` included which would work in simple use-cases. Which would make it change for v0.9.

What do you think?

**Comment by iNecas** [comment]
Nice catch with the `set_shutdown_hook`. I first thought that was something Java scpecific (as it wasn't happening on MRI). Definitely +1 to not do so automatically, especially not having different approaches in jruby vs. others.

Not having there the at_exit termination would definitely fix my testing issues

**Comment by jdantonio** [comment]
@pitr-ch @iNecas I believe that the `at_exit` hook should be created by default. To not do so will lead to a very bad experience for many users. The `java.util.concurrent` thread pools must be manually shut down within the application or the process will simply hang on exit, become completely unresponsive, and need manually killed by a user. Since we silently create the global thread pools without the user's knowledge we cannot expect our users to manually shut down these thread pools within their application. When a user manually creates a thread pool the danger is lessened since we can document the problem, but the failure condition is pretty severe. Assuming that our user's will always read the documentation is a big leap of faith.

If we add the `set_shutdown_hook` option and make the default `true` we'll protect the users who aren't deeply versed in concurrency while still giving the advanced users absolute control.

Rather than having the pure-Ruby thread pools ignore this flag, I think we should add the `at_exit` hook for all thread pools to make the behavior consistent. Adding the `at_exit` hook should have no practical effect when using pure-Ruby thread pools.

**Comment by jrochkind** [comment]
Under MRI, if you exit with running thread pools with current work in progress-- I think maybe it will actually kill threads and abandon work? 

In general, I think dealing with thread pools you created at program termination, is something you're going to have to do yourself, just like you do in Java. You have to decide whether you want to wait on them, or shut them down messily, or what. 

But I understand this isn't appropriate for the default built-in thread pools. I agree you need to come up with a way to handle them by default -- but do not think it should apply _by default_ to any thread pool created by client code (not the default core pools created by concurrrent-ruby itself). I think you are better off forcing the user to recognize the problem and decide how to deal with it.  

That is, I'd only default `set_shutdown_hook` to true on the built-in default thread pools, if you go that route.

**Comment by jdantonio** [comment]
@jrochkind Yes, under MRI when you exit the program the runtime will kill all the threads, regardless of what they are doing. You'll lose those tasks as well as any tasks on the queue that have not been started yet. And therein lies the problem. Under JRuby the thread pools will prevent the process from exiting unless they get properly shut down. The only way we can get that behavior on MRI would be to add an `at_exit` hook that prevents the application from closing.

If we want the shutdown behavior of *all* our thread pools (Java-backed and pure-Ruby) to be identical we *must* do something with `at_exit`. I don't see a way around that unless someone can figure something else out. Precisely *what* we do is the question.

If we make the default behavior of our thread pools be "no at_exit hook" then the Java-backed and pure-Ruby thread pools will behave differently. The Java thread pools will prevent the application from exiting but the pure-Ruby ones will not. If we want Java-like behavior from the pure-Ruby thread pools then we need to *add* an `at_exit` hook which causes the program to not exit unless all thread pools have been shut down. That would be a pretty intrusive change.

Sadly, I don't think we're going to find a perfect solution here. Every option seems to have a downside.

**Comment by iNecas** [comment]
I'm ok with both doing something at_exit to the default thread pools, or not, as far as it's consistent on all the implementations, ideally configurable. The problem for me was just, that it was working without issues on MRI, while I saw some strange errors on JRuby. From the adoption point of view, having the default shutdown makes sense.

**Comment by jrochkind** [comment]
Hmm, I see, yeah, thanks @jdantonio.  That does make me understand why the originaloriginal solution of an at_exit hook everywhere makes sense... but then the question is _why_ it was causing a problem as reported in this issue. Does anyone have a handle on that?

**Comment by rkday** [comment]
@jrochkind - I think it's causing a problem because minitest/autorun does its auto-running magic through another at_exit hook, and the two hooks conflict.

**Comment by jrochkind** [comment]
Aha, right. 

> If multiple handlers are registered, they are executed in reverse order of registration.

A workaround might be messing with when minitest vs concurrent-ruby gets loaded, to try to get concurrent-ruby's `at_exit` to run last.  But I realize that's not very satisfactory, I'm really just brainstorming here.  

But I wonder if the current behavior is really mostly right, and we just need a special way to disable/change it in the special circumstances where it is not, like minitest/autorun. Maybe a way to just disable all the at_exit behavior in some cases, but everything is still registered somewhere, where you can manually call something like `Concurrent.at_exit` yourself, perhaps in your test_helper.rb -- or maybe in most test cases, you don't even need the `at_exit` behavior, and can just disable it and not worry about some workers being killed in progress (only in MRI), cause you were testing and those guys are lost if the tests are finished anyway. 

Maybe it's good enough to just leave things as they are, but with a way to disable `at_exit` behavior entirely, which it could be recommended you do in your test_helper when using minitest/autorun (or other test autorunners using at_exit).

**Comment by jrochkind** [comment]
Wait, every single Executor adds it's _own_ `at_exit` hook when it's created? That's what it looks like looking at the code. 

It might make things more manageable (and performant?) if they added themselves to a queue or something, and there was a single global `at_exit` hook that killed everything in the queue. That'd make it easier to "turn off", and/or call yourself at program termination, "Okay, do all the stuff we're supposed to do before termination" 

Still kind of just brainstorming, I'm thinking of implications already making it more complicated. Sorry if I'm just being annoying.

**Comment by iNecas** [comment]
Just to make the discussion complete, we already do have such as option to enable disable that: https://github.com/ruby-concurrency/concurrent-ruby/blob/f104631243098a7e4a06b1b05181ae19c429679b/lib/concurrent/configuration.rb#L165 it only is not used in the Java thread pool

**Comment by jrochkind** [comment]
Thanks @iNecas that method, [an at_exit set at class load in configuration.rb](https://github.com/ruby-concurrency/concurrent-ruby/blob/f104631243098a7e4a06b1b05181ae19c429679b/lib/concurrent/configuration.rb#L165) --  which can be turned on or off with configuration -- seems to operate only on the global threadpools, and seems to call `shutdown` but *not* `wait_for_termination`. If I understand right, on JRuby that will effectively make the interpreter wait for all in-progress-or-queued work to complete, because the interpreter won't exit on running threads. But on MRI, it will do almost nothing, as you'll tell the global thread to shutdown, but then the interpreter will exit anyway before it's finished. 

And I'm not sure if there is a purpose for `executor.respond_to?(:shutdown) ... elsif executor.respond_to?(:kill)`. Are there any executors in current version that don't respond to shutdown? I don't think so. Definitely not any of the global executors?

Meanwhile, we have *also* [set_shutdown_hook](https://github.com/ruby-concurrency/concurrent-ruby/search?q=set_shutdown_hook&type=Code&utf8=%E2%9C%93) pointed out by @pitr-ch , which *seems* to be called for *all* executors, and *also* sets an at_exit hook.  I don't see any reason this wouldn't be called for the global executors too, it seems to apply to any executors on initialization (although maybe only in the Java implementations? Maybe it's meant to make Jruby work like MRI, where in-progress-or-queued work will be abandoned at exit, instead of preventing program termination?)  

Does that mean that the *global* executors now have *two* at_exit hooks operating on them (at least in Java?), the one set in configuraton.rb for just global executors, *and* the one that will be set by the individual executor `set_shutdown_hook`?  The `set_shutdown_hook` one does not try to shutdown, but just `kill`s (cancelling any in-progress or queued work), so if that happens before the global executor at_exit hook @iNecas points out, it's just pre-empted it anyway. 

It seems like the existing setup might be a bit of a muddle, and needs to be unified, at one place in the code, at least?

**Comment by jrochkind** [comment]
For comparison, it looks to me like celluloid also [installs an at_exit handler](https://github.com/celluloid/celluloid/search?utf8=%E2%9C%93&q=at_exit), but provides a file that you can [require 'celluloid/test'](https://github.com/celluloid/celluloid/blob/c31d1e9277db888f472bc95ff1cf4ae516f409e0/lib/celluloid/test.rb) which will disable the at_exit handler -- it actually simply sets `$CELLULOID_TEST = true`,  which [prevents the registration of the shutdown handler](https://github.com/celluloid/celluloid/blob/c31d1e9277db888f472bc95ff1cf4ae516f409e0/lib/celluloid.rb#L516).   

It's unclear to me [if this functionality](https://github.com/celluloid/celluloid/wiki/Testing) was intended for dealing with things like minitest/autorun (or rspec/autorun which works the same way, but which people use less often) -- or if it was intended for testing Celluloid itself.  

If you used it for minitest/autorun, under Jruby I think it might prevent your process from ever shutting down, as Celluloid won't shut down the thread pools, you probably won't do it manually either, and that will keep the process from shutting down? 

I think the concurrent-ruby stuff probably does need to be cleaned up and unified -- a separate at_exit handler created on initialization of every executor seems like a bad idea no matter what. As does the fact that individual concrete executor base classes are responsible for doing this. And then the conflict between those, and the global executor at_exit in configuration.rb. 

I think there probably needs to be a single `at_exit`, which shuts down all executors, which have registered themselves (and de-register on their own manual shutdown?). This will make it easier to see what's going on and avoid conflicts, and easier to change the behavior of `at_exit` shutdow, because it's been DRY'd.  Change it by configuration -- for instance to disable it entirely. Or change it if concurrent-ruby devs change their mind what the default behavior actually should be (should it kill, or wait for work to be completed? Or wait with a timeout?). 

But default behavior probably needs to remain the same, with an at_exit. if it conflicts with minitest/autorun, the solution is making sure you're "require's" are in the right order to prevent the conflict -- you need minitest/autorun's `at_exit` to run _before_ ruby-concurrents.  (But there's probably a reason rspec kind of leans against using at_exit based autorun, and running with `rspec` command instead).

**Comment by jdantonio** [comment]
Thank you everyone for your thoughts and research. I think we have an understanding of the situation and enough information to craft a solution. As soon as I have an opportunity I will post a lengthy summary/explanation of the problem, the history of our executors that lead to the current (somewhat inconsistent) implementation, and my thoughts on how we should proceed. I just got to work so it will take some time for me to put this all together, but I promise to post sometime today.

**Comment by jdantonio** [comment]
## Background

The original thread pool implementations were very simple and were written in pure-Ruby (no JRuby wrappers around java.util.concurrent). There were four thread pools: `FixedThreadPool`, `CachedThreadPool`, `NullThreadPool` (now called `PerThreadExecutor`), and `EventMachineDeferProxy` (long-since deprecated and removed). The *minimum* API for thread pools at that time was a `#post` method. There was no `Executor` superclass and no mixin modules. There was only one global thread pool (the default was a `NullThreadPool`). The global thread pool could be changed by the user via the `$GLOBAL_THREAD_POOL` global variable.

I don't remember which came first--the global configuration class or the Java thread pools--so I'll discuss them separately.

The Java thread pools were originally created as variations of the Ruby `FixedThreadPool` and `CachedThreadPool` classes. The initial implementation did not include the current hierarchy with `Executor`, `ThreadPoolExecutor`, etc. All of that stuff came later over a series of PRs which ensured backward compatibility. When I first created the JRuby thread pool classes that wrap java.util.concurrent I discovered the JVM issue I [explained above](https://github.com/ruby-concurrency/concurrent-ruby/issues/248#issuecomment-74506961). The JVM simply refuses to shut down if there are running thread pools, even when all tasks on the thread pools have completed. That is when I created the `at_exit` handler (via the `set_shutdown_hook` method) within the JRuby thread pools. This was our first foray into JRuby-specific classes so I had yet to established my "implementations must have the same behavior" rule. Subsequently I did *not* create `at_exit` handlers for the Ruby thread pools (an oversight, in retrospect).

I've never liked the use of global variables (such as `$GLOBAL_THREAD_POOL`) so another early refactor was to create the current global `Configuration` class. The initial implementation was much simpler, created only one global thread pool, and did not use `Delay` to lazy-load that thread pool. I created the `at_exit` handler in the earliest versions of `Configuration` as a way to ensure that we shut down the global thread pool(s) in an orderly way. To be perfectly honest I gave no thought to the interaction of the `at_exit` handler in the global config with the `at_exit` handler in each JRuby thread pool. It wasn't until we merged the nascent `Actor` framework into this gem that our global thread pools started to be become a problem. Our `Actor` framework is the most complex part of this gem and we had to evolve our global configuration along with it. That's when we created two global thread pools, added the `Delay` objects, added a global logger, created the `auto_terminate` feature, etc. These changes happened independent of the extensive thread pool refactoring and evolution since those changes were backward compatible.

As has been noted in the comments, the `auto_terminate` feature of our global configuration allows the user to prevent the `at_exit` handlers from being created against the global thread pools. This feature does *not* affect the creation of the `at_exit` handlers from being created during the construction of each individual Java-backed thread pools. There is currently no way for a user to prevent each Java-backed thread pool from creating an `at_exit` handler.

The various implementations of our global configuration have always allowed the global thread pool(s) to be replaced. Since the initial thread pools only required a `#post` method the original `at_exit` handler in the `Configuration` class needed to check for the existence of the `#shutdown` and `#kill` methods. That may not be necessary any more given the current executor implementations.

The current class/module hierarchy for our executors was created over a very grueling set of PRs that also involved major changes to the tests (by creating several example groups). The main goal was to align our thread pools more with java.util.concurrent, most especially by creating `ThreadPoolExecutor`. Every PR remained backward compatible. Once I got everything working I pretty much walked away from thread pools for a while. I believe there is still room to refactor that code, especially with respect to the class/module hierarchy.

## Problem

The problem this issue was created to address is a conflict between our `at_exit` handlers and the use of `at_exit` within [autotest](https://github.com/seattlerb/zentest/blob/master/lib/zentest.rb#L37). A couple of well-know Ruby applications, including autotest and [Goliath](https://github.com/postrank-labs/goliath/blob/master/lib/goliath/application.rb#L127) use `at_exit` as a "clever" way to force their application to run. The entire application runs in an `at_exit` block. I am vehemently opposed to this practice. [Ruby's at_exit](http://ruby-doc.org//core-2.2.0/Kernel.html#method-i-at_exit) is intended for cleanup "when the program exits." It is *not* intended to be a clever way to run your application. The problem is that `at_exit` as used by autotest is incompatible with the use of `at_exit` in our thread pools.

The core of the issue is this: "If multiple handlers are registered, they are executed in reverse order of registration."

When a user of our gem *also* uses autotest, the `at_exit` handler for autotest is registered first, followed by ours. This causes our handlers to *run* first, shutting down our thread pools. The *correct* solution, in my opinion, is for autotest to *not* use `at_exit` in such a hackish and disruptive way. But I have no illusions--that isn't going to happen. This means that we need to find a way for our thread pools to play nicely with autotest,

We must *also* remember that the behavior of the JVM can't be changed, either. The JVM will not shut down with running thread pools. So whatever shenanigans we implement to make autotest happy must *also* play nicely with the JVM.

## Goals

Our solution must support the following goals:

* **Simple for the average Rubyist:** The primary goal of this gem has always been to make concurrency as simple as reasonably possible for the average Rubyist who is not well-versed in concurrency. A simple script that [creates a handful of futures then aggregates the results](http://rockyj.in/2014/05/27/concurrent_ruby_jruby.html) is a reasonable base case. A user writing a script like that shouldn't need to interact with the global config and should expect the JVM to shutdown properly when the script is done.
* **Provide advanced options for advanced users:** As we've evolved we've added many new features that give greater control to users with advanced concurrency knowledge. It's reasonable for an advanced user to want greater control of our `at_exit` handlers.
* **Give priority to the average Rubyist:** Whenever we add complexity we always make the default setup favor the simple use case of the average Rubyist.
* **Be consistent across all implementations:** Whenever we have platform-specific implementations we take steps to ensure that behavior is consistent across all implementations, even if this means adding features to the pure-Ruby implementations.
* **Allow low-level abstractions to be used independently:** One of the strengths of this gem is that our low-level abstractions are loosely coupled and are exposed independently. A use who only needs atomic objects or channels or thread pools can use them directly without creating unnecessary overhead within their application. That's why we lazy-load our global thread pools. Our solution to this problem should not burden these users with additional configuration or overhead.
* **Be nice to the users of autotest:** It is the fault of the user that autotest runs the entire application within an `at_exit` handler. We should find a simple way to configure the gem that will be compatible with autotest.

## Proposed Solution

With all this in mind, here is what I suggest we do:

* Add a new option to the constructor of the Java-backed thread pools that controls the creation of the `at_exit` handler
  - The default will be to *enable* the `at_exit` handler
  - This will support the "simple and easy" base case
  - And will ensure backward compatibility with the current release
* Create an `auto_shutdown?` (or similarly named) predicate method to our Java-backed thread pools that indicates whether the `at_exit` handler has been set
* Update the pure-Ruby thread pools with the *exact* same `at_exit` behavior as the Java-backed thread pools
  - It was an oversight that our pure-Ruby thread pools are not 100% compatible with our Java-backed thread pools
  - However, we need to make sure that this behavior is backward compatible with the current release
* Update the global config to ensure only one `at_exit` handler ever exists for each global thread pool
  - Our default global thread pools should be created without their own `at_exit` handlers
  - Let the `at_exit` handlers created within the global config handle the global thread pools
* Update the behavior of the global configuration `auto_terminate` feature
  - Remove unnecessary `respond_to?` checks
  - Ensure proper handing of `at_exit` handler creation for all thread pools automatically created within the global config
  - Ensure proper handling of `at_exit` when the user explicitly sets a global thread pool
* Create a global "autotest friendly" configuration option
  - I'm not 100% certain what this will look like yet
  - But it should include a `require 'concurrent/autotest'` feature for simple setup
  - It should prevent any `at_exit` handlers from being created for *all* thread pools, regardless of how they are created
  - Yet it should also ensure that the JVM should not hang because the thread pools are not shutdown
  - And it should be as unobtrusive as possible (I hope to avoid coupling thread pools to some global registry if possible)
* Document, document, document

All thoughts, suggestions, feedback, and comments are welcome...

**Comment by jrochkind** [comment]
I like it. 

But would you consider registering each thread pool meant for auto-shutdown in a (thread-safe) collection, with a single `at_exit` handler that shuts down all such registered pools?

You say you hope to prevent a global registry....  but you essentially have a global registry now, the global registry is just the ruby interpreter's list of `at_exit` hooks. Which is not a very handy "global registry" (yeah, I'm stretching the semantics to call it that at all), you can't do much with it, it's hard to reason about, it's hard to be sure you are finding all the relevant source code for it or that every at_exit code does the same thing, it's hard to change it if you decide "Oh, actually, I want to `shutdown && wait_for_termination` on all those pools, not `kill`" (or vice versa, either globally or per-pool configuration), etc. 

A single at_exit handler operating on a global registry just _feels_ a lot cleaner to me, a lot easier to reason about, maintain, and alter. What are the downsides?  (And are there any downsides to having potentially dozens (or more?) separate `at_exit` hooks registered if you have dozens+ thread pools, which seems possible?)

**Comment by jrochkind** [comment]
And is it possible to be more parsimonious, and not have entirely separate logic path for the global threadpools?   Rather than:

> * Our default global thread pools should be created without their own at_exit handlers
> * Let the at_exit handlers created within the global config handle the global thread pools

Why can't the global thread pools be handled the same as all other thread pools -- if thread pools may or may not be autoshutdown?, then of course the global thread pools would be, but why do they need a separate path to treat them special?

Let's say I realize I should create my Futures off a custom thread pool, because I need certain particular configuration (caller_runs policy, queue length, min/max threads).  Now I'm no longer using the global thread pool... to make sure things work right, I need to understand not only the standard applies-to-any-pool hypothetical autoshutdown behavior, but I also have to understand the special behavior for global thread pools, to consider if I want to duplicate it for my custom thread pool so it behaves like the global ones (and then I'd have to duplicate it, since the special global behavior is only for the registered global threadpools). 

Why not have just one kind of thread pool auto shutdown mechanism, used for all thread pools, and configure the global thread pools with appropriate parameters so the general auto shutdown mechanism will handle them as desired?

**Comment by jdantonio** [comment]
May main concerns with a global registry for thread pools are bloat and complexity.

One of the strengths of this gem is that we have very loose coupling between most components. A user can create a single thread pool and that's what they get--a single thread pool. You don't get a bunch of hidden scaffolding created to support it. And this is the case for most of the gem. (The main exception being actors, which by their very nature need that scaffolding).

In my experience, once you start building global support objects, that stuff starts to grow. In our 0.5 release our global configuration was a single thread pool assigned to a single global variable. That was it. My initial implementation of the Configuration class had only [a dozen lines of code](https://github.com/ruby-concurrency/concurrent-ruby/blob/6ec154decaaa45aa7127a6bfff99b2884f72f5df/lib/concurrent/configuration.rb) or so. With the 0.8 release it's grown to [over 80 lines of code](https://github.com/ruby-concurrency/concurrent-ruby/blob/v0.8.0/lib/concurrent/configuration.rb).

How far do we let the global configuration grow before we decide it's gotten too big?

An at_exit hook created within each thread pool encapsulates the shutdown behavior within the class itself. Moving the shutdown behavior to a global registry couples our thread pools to the registry and it breaks the encapsulation of the class. Thread pools are no longer loosely coupled, highly cohesive, object oriented abstractions. They are now part of a framework, and that's inherently more complex.

In the example you give, where you decide to create a custom thread pool for your futures, you simply create the thread pool and use it. The behavior of the global thread pools is irrelevant to you. The global thread pools may never even be created since you're using your own thread pool. If one or more of the global thread pools *does* get created (perhaps you create a timer somewhere else in your code) then the default behavior--shutting down the global thread pool on exit--is *exactly* what you want from a thread pool that you didn't know was being created.

And that's the whole point: Users of this gem shouldn't know or care about the global thread pools. They should behave in a safe, reliable way that just gets the job done. And for the most part that's what happens now--except when we encounter a misbehaving program like autotest. So we want to create a simple, reliable way for our thread pools to work under autotest so that our users don't have to think about them there, either.

**Comment by jrochkind** [comment]
Okay, I see your objection to global registry, although I think there are pro's and con's both ways, your position is not unreasonable. :) 

Is there a reason the global thread pools need their own separate `at_exit` handler though, can they just use the same mechanism for auto-shutdown as every other threadpool?   Is there a reason the global pools need to be handled specially? 

(I think I saw in another issue someone, you perhaps, suggesting that changing the global pools ought to be discouraged or impossible; if you want a different pool you can create one and use it wherever you like in place of the global, but you shouldn't change the globals. I agree with this (otherwise third party code using ruby-concurrent can't be confident the global pools will work as expected), and if followed through it might make things simpler for global autoshutdown). 

It would be good, if possible, to put that decoupled one-at_exit-handler-per-pool code in a base class somewhere, not need to duplicate it in every sub-class. And even more ideally, same code used for global pools and other pools.  That would make it easier to ensure consistency, make changes/fix bugs, and for a developer to find the code (in one place) and see what's going on. 

I am not sure if I am following all the implications right -- if I create my own thread pool to use, say, with  Futures -- by default, will I have to do cleanup myself to ensure clean shutdown on all platforms (including jvm), or by default will it be taken care of for me?  Ideally, using a custom threadpool with (say) futures should not require special custom cleanup that the globals don't, you should be able to just create the pool and pass it to the Future. 

Is there any reason to be concerned about "too many separate at_exit handlers" if you create a lot of pools?

**Comment by jrochkind** [comment]
It also might be nice if you could configure an individual pool with it's at_exit shutdown behavior -- you might want a pool to do a `shutdown && wait_for_termination(some_timeout)` at exist, instead of `kill`.

I don't think this feature needs to or should be in the first release of a refactored at_exit handling. Just bring it up as potentially an example use case to mentally test against the architecture to see if the architecture would make it easy to add in the future if desired.

**Comment by jdantonio** [comment]
@iNecas @rkday Can you please provide me with example code that triggers this exception along with instructions on how to reproduce the error? I don't use minitest/autotest myself. I've been working on these (and other) updates to our global thread pools and I'd like to start exploring the minitest/autotest issue.

**Comment by rkday** [comment]
@jdantonio - this should reproduce it (RejectedExecutionError when running under MRI 2.1.5 or JRuby 1.7.16.1). Switching the order of the `require` statements (and thus the order of at_exit hooks) fixes it for me - so this is slightly different from @iNecas 's report where some code worked under MRI but not under JRuby.

```
$ cat example.rb 
require 'minitest/autorun'
require 'concurrent'

class TestExample < MiniTest::Unit::TestCase
  def test_future
    i = 0
    Concurrent::Future.execute {i = 10}
    sleep 2
    assert_equal i, 10
  end
end
```

**Comment by jdantonio** [comment]
@rkday I've verified this issue on jruby-1.7.18. The problem does *not* appear to exist on the `fix/thread-pool-shutdown` branch being used for PR #255. Please confirm if you have a chance.

The problem appears to have been related to an `Actor` bug which I fixed in commit 3f2df1c. Our configuration uses `LazyReference` objects (previously `Delay`, changed in 7fb5a0d). These prevent our global thread pools from being created unless actually used. The bug was an incorrect use of Delay. The Delay used to [create the root actor](https://github.com/ruby-concurrency/concurrent-ruby/commit/3f2df1c329ab3171738ace032256c39897df88f4#diff-bd08c2b99f1aa6874a61fb7a8c6f5dd6R42) was running on one of the global thread pools rather than on an `ImmediateExecutor`. Subsequently, the Delay warmed up the thread pool even though the root actor was not being created. This, in turn, caused the global thread pool to *always* be created immediately upon `require 'concurrent'`. And this, in turn, caused the `at_exit` handler to be created prematurely.

The fix in 3f2df1c properly uses `LazyReference` to lazy-load the root actor and the global thread pools. This ensures that `minitest/autorun` creates its `at_exit` handler first, regardless of the order of includes.

**Comment by jdantonio** [comment]
@iNecas If you have time can you please check the `fix/thread-pool-shutdown` branch and see if the change I mentioned above fixes your issue?

**Comment by iNecas** [comment]
I confirm that my issue is not reproduced on the new branch

**Comment by jdantonio** [comment]
@iNecas :-) Thank you!

**Comment by iNecas** [comment]
THANK YOU! The cr team just rocks

**Comment by jdantonio** [comment]
Addressed in PR #255.
